### PR TITLE
[SYSTEMML-445] Refactoring to avoid potential memory leaks

### PR DESCRIPTION
**Pull request by niketanpansare**
- Removed tensor descriptors from GPUObject
- Created closeable LibMatrixCuDNNPoolingDescriptors class to manage the data structures required by maxpooling
- Enabled JCuda exceptions to catch CUDA errors eagerly
- Added debugging messages in eviction logic. The printing of these messages is guarded to avoid additional overhead
- Removed unused batch normalization methods from LibMatrixCuDNN

@nakul02 can you please review this PR ?

### CODE REVIEW

File: src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUContext.java [hunk 1]
```diff
@@ -242,6 +243,10 @@ private void initializeCudaLibraryHandles() throws DMLRuntimeException {
 		if (kernels == null) {
 			kernels = new JCudaKernels();
 		}
+		// The CUDA binding methods will throw a CudaException if the CUDA function did not return cudaError.cudaSuccess.
+		JCuda.setExceptionsEnabled(true);
```
Comment (nakul02) [inline src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUContext.java]: You don't need to set `JCuda.setExceptionsEnabled` & `JCudaDriver.setExceptionsEnabled`
They are already done in `GPUContextPool`

File: src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUContext.java [hunk 2]
```diff
@@ -401,21 +406,26 @@ public void cudaFreeHelper(String instructionName, final Pointer toFree) {
...
 		if (eager) {
 			if (LOG.isTraceEnabled()) {
-				LOG.trace("GPU : eagerly freeing cuda memory [ " + toFree + " ] for instruction " + instructionName
+				LOG.trace("GPU : eagerly freeing cuda memory [ " + toFree + " ] of size " + size + " for instruction " + instructionName
 						+ " on " + this);
 			}
 			if (DMLScript.STATISTICS)
 				t0 = System.nanoTime();
 			cudaFree(toFree);
+			jcuda.runtime.JCuda.cudaDeviceSynchronize();
```
Comment (nakul02) [inline src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUContext.java]: `cudaFree` and `cudaMalloc` are synchronous operations. You don't need to put this `cudaDeviceSynchronize` here. (if in fact, that is your intent).

File: src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUObject.java [hunk 3]
```diff
@@ -132,11 +114,6 @@
 	public Object clone() {
 		GPUObject me = this;
 		GPUObject that = new GPUObject(me.gpuContext, me.mat);
-		if (me.tensorShape != null) {
-			that.tensorShape = new int[me.tensorShape.length];
-			System.arraycopy(me.tensorShape, 0, that.tensorShape, 0, me.tensorShape.length);
-			that.allocateTensorDescriptor(me.tensorShape[0], me.tensorShape[1], me.tensorShape[2], me.tensorShape[3]);
-		}
 		that.dirty = me.dirty;
 		// TODO Nakul: Should the locks be cloned here ?
```
Comment (nakul02) [inline src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUObject.java]: I think you can remove this TODO :)

File: src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixCuDNN.java [hunk 4]
```diff
@@ -506,41 +489,29 @@ public static void maxpooling(GPUContext gCtx, String instName, MatrixObject ima
...
 			int S, int pad_h, int pad_w, int stride_h, int stride_w, int P,
 			int Q) throws DMLRuntimeException {
 		if(LOG.isTraceEnabled()) {
 			LOG.trace("GPU : performMaxpooling" + ", GPUContext=" + gCtx);
 		}
 
-		cudnnPoolingDescriptor poolingDesc = null;
-
-		try {
+		try(LibMatrixCuDNNPoolingDescriptors desc =
```
Comment (nakul02) [inline src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixCuDNN.java]: I like this. You may want to put the `MISC_TIMER_CUDNN_CLEANUP` timer in the `LibMatrixCuDNNPoolingDescriptors` class.

**Comment by niketanpansare** [comment]
@nakul02 addressed your comments.

**Comment by nakul02** [comment]
LGTM 👍

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2173/

**Pull request closed without merging**